### PR TITLE
New Smart Fridge sprites

### PR DESCRIPTION
**Pull request by MTandi**
## About The Pull Request

Resprites smart fridges. Inspired by the smart fridge by @Krysonism.

dreamseeker_YHGPSCCEs3

dreamseeker_1HJFZwQbjR

Versions:

image

## Why It's Good For The Game

It looks nice.

## Changelog

:cl: @Krysonism, @Imaginos16, @MTandi, @san7890 
image: updated smart fridge sprites for all variants
fix: smart fridges no longer overlap with walls, but they can be welded down to a floor and mapspawn welded
/:cl:

### CODE REVIEW

**Comment by Overuseddart** [comment]
Wait can you see what items are in the fridge from the sprite?

**Comment by MTandi** [comment]
> Wait can you see what items are in the fridge from the sprite?

No, it just tricks you thinking that way. It's just a set of 3 sprites of every of 8 types of fridges.

**Comment by dragomagol** [comment]
Looks fantastic, although I expect we'll need some mapping changes following this because there are some placed in walls.

**Comment by MTandi** [comment]
> Looks fantastic, although I expect we'll need some mapping changes following this because there are some placed in walls.

You can see on the second gif that it's not really that noticeable. But sure we may need to change them to glass or something.

**Comment by san7890** [comment]
image

this is quite noticeable to my eyes.

**Comment by CoiledLamb** [comment]
wait, the item overlays show even if the machine is broken? does that not lead to a bunch of weird edge cases where the overlay'd item is floating over nothing?

**Comment by MTandi** [comment]
> wait, the item overlays show even if the machine is broken? does that not lead to a bunch of weird edge cases where the overlay'd item is floating over nothing?

In the broken state only the top shelf is partially skewed, and it doesn't look bad even when the fridge is full. You can still extract contents from it with crowbar when it is broken.

**Comment by CoiledLamb** [comment]
> > wait, the item overlays show even if the machine is broken? does that not lead to a bunch of weird edge cases where the overlay'd item is floating over nothing?
> 
> In the broken state only the top shelf is partially skewed, and it doesn't look bad even when the fridge is full. You can still extract contents from it with crowbar when it is broken.

would be nice to get pictures of this...
i get that if the fridge is destroyed the items inside don't disappear, but i feel like not rendering the item overlay when the machine is broken is internally consistent with most other types of broken displays in the game. vending.dmi might be a good frame of reference for that

**Comment by MTandi** [comment]
image

**Review by san7890 — CHANGES_REQUESTED** [review]
yeah the more i look at it the less I like the wall colliding with the smart fridge. i would make a maplint against closed turfs and smartfridges being on the same turf (https://github.com/tgstation/tgstation/tree/master/tools/maplint) and then put out any fires mapping-wise.

**Review by san7890 — CHANGES_REQUESTED** [review]
There's two things still left to do:

First, per @LemonInTheDark, smart-fridges (since we're replacing the walls) should be able to block atmospherics when fully intact (and preferably be free-to-atmospherics if broken open). He wrote a bit about this in Discord and tagged you with it, but I'm pasting a screengrab of it below for posterity:

image
image


Also, since they're pretty easy to disassemble to gain entrance to a room now, I think we should make it such that the fridge is bolted in place (and is not deconstructible while the bolts are active) which is controlled by a wire that must be multitooled. This is simply to make it such that you can't easily gain access to pharmacy or something by simply unwrenching a fridge- there needs to be more work involved. We could make it more secure by making it so that the panel would have to be welded through (to mimick that action that had to occur when they were still built on walls) but that's up to you.

let me know if you need any help with implementing this, sorry for scope-creeping your PR though all of these reviews.

**Comment by MTandi** [comment]
> Also, since they're pretty easy to disassemble to gain entrance to a room now, I think we should make it such that the fridge is bolted in place (and is not deconstructible while the bolts are active) which is controlled by a wire that must be multitooled. This is simply to make it such that you can't easily gain access to pharmacy or something by simply unwrenching a fridge- there needs to be more work involved. We could make it more secure by making it so that the panel would have to be welded through (to mimick that action that had to occur when they were still built on walls) but that's up to you.

Since this is also true for the ORM, I suggest to do this in a separate PR and make it through a helper, because drying rack is a subtype of smartfridge and the solution here would be ugly.

**Comment by san7890** [comment]
> Since this is also true for the ORM, I suggest to do this in a separate PR and make it through a helper, as fridges are not always used inside walls. There is also drying rack which is a subtype of smartfridge and the solution here would be ugly.

okay

**Review by san7890 — APPROVED** [review]

**Comment by MTandi** [comment]
@san7890 I made the ATMOS changes as suggested, but I'm not sure it works fine as I never debugged LINDA. Is there a way to check that air is properly changed when you initialize and move the machine? That it does not delete the air on a turf it moves to for example.

File: code/modules/food_and_drinks/machinery/smartfridge.dm [hunk 1]
```diff
@@ -34,6 +40,10 @@
 			for(var/i in 1 to amount)
 				load(new typekey(src))
 
+/obj/machinery/smartfridge/Moved(atom/old_loc, movement_dir, forced, list/old_locs, momentum_change)
+	. = ..()
+	air_update_turf()
```
Comment (LemonInTheDark) [inline code/modules/food_and_drinks/machinery/smartfridge.dm:45]: I fed you bad info on the proc to override, we have a helper for this sorta thing
We need to update both the new turf and the old one, which this doesn't do
```suggestion
/obj/machinery/smartfridge/Move(atom/newloc, direct, glide_size_override, update_dir)
	var/turf/old_loc = loc
	. = ..()
	move_update_air(old_loc)
```

Comment (MTandi) [reply code/modules/food_and_drinks/machinery/smartfridge.dm]: I made it update air only on wrench/unwrench. It is working right now with the recent change.

File: code/modules/food_and_drinks/machinery/smartfridge.dm [hunk 2]
```diff
@@ -34,6 +40,12 @@
 			for(var/i in 1 to amount)
 				load(new typekey(src))
 
+
+/obj/machinery/smartfridge/set_anchored(anchorvalue)
+	. = ..()
+	can_atmos_pass = anchorvalue ? ATMOS_PASS_NO : ATMOS_PASS_YES
+	air_update_turf(TRUE, anchorvalue)
```
Comment (LemonInTheDark) [inline code/modules/food_and_drinks/machinery/smartfridge.dm:47]: atoms can move when anchored, I'd still like that Move() override

**Comment by MTandi** [comment]
@san7890 I made it work like emitters right now. To dismantle it, you need to unweld it first and only then unwrench, so the amount of effort is the same as dismantling a wall.

**Comment by san7890** [comment]
> In fact, this should be the default machine behavior IMO. You should be able to weld down any subtype of machine. This will solve the ORM issue too and just provide more options to mapmakers by making other machines secured (Like chem dispensers 😈). But this is outside of scope of this PR.

The consequences of this aren't really measurable in this PR and I'm uncertain how I feel about this being applied to more machinery. Thanks for doing it here though, it was surely needed.

File: code/modules/food_and_drinks/machinery/smartfridge.dm [hunk 3]
```diff
@@ -34,39 +44,153 @@
...
+	if(!anchored)
+		to_chat(user, span_warning("[src] needs to be wrenched to the floor!"))
+		return TRUE
+	if(!tool.tool_start_check(user, amount=1))
+		return TRUE
+	user.visible_message(span_notice("[user.name] starts to weld the [name] to the floor."), \
+		span_notice("You start to weld [src] to the floor..."), \
+		span_hear("You hear welding."))
+	if(!tool.use_tool(src, user, delay=100, amount=1, volume=100))
+		return FALSE
```
Comment (san7890) [inline code/modules/food_and_drinks/machinery/smartfridge.dm:89]:
```suggestion
	if(!tool.use_tool(src, user, delay=100, amount=1, volume=100))
		balloon_alert(user, "cancelled!")
		return FALSE
```

File: code/modules/food_and_drinks/machinery/smartfridge.dm [hunk 4]
```diff
@@ -34,39 +44,153 @@
...
+		to_chat(user, span_notice("You cut [src] free from the floor."))
+		return TRUE
+	if(!anchored)
+		to_chat(user, span_warning("[src] needs to be wrenched to the floor!"))
+		return TRUE
+	if(!tool.tool_start_check(user, amount=1))
+		return TRUE
+	user.visible_message(span_notice("[user.name] starts to weld the [name] to the floor."), \
+		span_notice("You start to weld [src] to the floor..."), \
+		span_hear("You hear welding."))
```
Comment (san7890) [inline code/modules/food_and_drinks/machinery/smartfridge.dm:87]:
```suggestion
	user.visible_message(
		span_notice("[user.name] starts to weld the [name] to the floor."),
		span_notice("You start to weld [src] to the floor..."),
		span_hear("You hear welding."),
	)
```

native procs (as in, those without macro wrappers like `AddComponent()`) don't need backslashes in current DM and can (and should) be multi-lined like any list.

File: code/modules/food_and_drinks/machinery/smartfridge.dm [hunk 5]
```diff
@@ -34,39 +44,153 @@
...
+/obj/machinery/smartfridge/welder_act(mob/living/user, obj/item/tool)
+	..()
+	if(istype(src, /obj/machinery/smartfridge/drying_rack))
+		return FALSE
+	if(welded_down)
+		if(!tool.tool_start_check(user, amount=1))
+			return TRUE
+		user.visible_message(span_notice("[user.name] starts to cut the [name] free from the floor."), \
+			span_notice("You start to cut [src] free from the floor..."), \
+			span_hear("You hear welding."))
```
Comment (san7890) [inline code/modules/food_and_drinks/machinery/smartfridge.dm:74]:
```suggestion
		user.visible_message(
			span_notice("[user.name] starts to cut the [name] free from the floor."),
			span_notice("You start to cut [src] free from the floor..."),
			span_hear("You hear welding."),
		)
```

File: code/modules/food_and_drinks/machinery/smartfridge.dm [hunk 6]
```diff
@@ -34,39 +44,153 @@
...
+	if(machine_stat & BROKEN)
+		if(!tool.tool_start_check(user, amount=0))
+			return FALSE
+		user.visible_message("<span class='notice'>[user] is repairing [src].</span>", \
+						"<span class='notice'>You begin repairing [src]...</span>", \
+						"<span class='hear'>You hear welding.</span>")
+		if(tool.use_tool(src, user, delay=40, volume=50))
+			if(!(machine_stat & BROKEN))
+				return FALSE
+			to_chat(user, "<span class='notice'>You repair [src].</span>")
```
Comment (san7890) [inline code/modules/food_and_drinks/machinery/smartfridge.dm:107]:
```suggestion
			balloon_alert(user, "repaired")
```

File: code/modules/food_and_drinks/machinery/smartfridge.dm [hunk 7]
```diff
@@ -34,39 +44,153 @@
...
+/obj/machinery/smartfridge/welder_act_secondary(mob/living/user, obj/item/tool)
+	. = ..()
+	if(istype(src, /obj/machinery/smartfridge/drying_rack))
+		return FALSE
+	if(machine_stat & BROKEN)
+		if(!tool.tool_start_check(user, amount=0))
+			return FALSE
+		user.visible_message("<span class='notice'>[user] is repairing [src].</span>", \
+						"<span class='notice'>You begin repairing [src]...</span>", \
+						"<span class='hear'>You hear welding.</span>")
```
Comment (san7890) [inline code/modules/food_and_drinks/machinery/smartfridge.dm:103]:
```suggestion
		user.visible_message(
			span_notice("[user] is repairing [src]."),
			span_notice("You begin repairing [src]..."),
			span_hear("You hear welding."),
		)
```

File: code/modules/food_and_drinks/machinery/smartfridge.dm [hunk 8]
```diff
@@ -34,39 +44,153 @@
...
+		if(tool.use_tool(src, user, delay=40, volume=50))
+			if(!(machine_stat & BROKEN))
+				return FALSE
+			to_chat(user, "<span class='notice'>You repair [src].</span>")
+			atom_integrity = max_integrity
+			set_machine_stat(machine_stat & ~BROKEN)
+			update_icon()
+			return TRUE
+	else
+		to_chat(user, "<span class='notice'>[src] does not need repairs.</span>")
```
Comment (san7890) [inline code/modules/food_and_drinks/machinery/smartfridge.dm:113]:
```suggestion
		balloon_alert(user, "no repair needed!")
```

if you have a better short message in mind use that instead.

**Review by san7890 — CHANGES_REQUESTED** [review]
bunch of UX stuff/formatting nits that should be addressed now

also merge master whenever you get the chance to ensure we don't skew on the maplint or something like that

File: code/modules/food_and_drinks/machinery/smartfridge.dm (hunk 7 above)
Comment (san7890) [reply code/modules/food_and_drinks/machinery/smartfridge.dm:103]: unsure why you didn't use the span macros here but you should pretty much always use those

Comment (MTandi) [reply code/modules/food_and_drinks/machinery/smartfridge.dm:108]: It's mostly copypasta from emitter, I just checked that the logic is correct without refactoring.

**Pull request merged**